### PR TITLE
MDEV-16429: Assertion `!table || (!table->read_set || bitmap_is_set(t…

### PR DESCRIPTION
**Pull request by FooBarrior**
…able->read_set, field_index))' fails upon attempt to update virtual column on partitioned versioned table

* update keyread state of underlying handlers recursively

test suites run: main,parts,versioning

### CODE REVIEW

File: sql/ha_partition.h [hunk 1]
```diff
@@ -864,6 +864,7 @@ class ha_partition :public handler
                                           **block_table,
                                           handler *file, uint *n);
   static const uint NO_CURRENT_PART_ID= NOT_A_PARTITION_ID;
+  int loop_partitions(int f(handler *, void *), void *param);
```
Comment (midenok) [inline sql/ha_partition.h]: Define type for callback.

File: sql/ha_partition.cc [hunk 2]
```diff
@@ -9102,14 +9131,40 @@ int ha_partition::loop_extra(enum ha_extra_function operation)
...
+*/
+
+int ha_partition::loop_extra(enum ha_extra_function operation)
+{
+  int result;
+  DBUG_ENTER("ha_partition::loop_extra()");
+
+  result= loop_partitions(extra_cb, &operation);
+  DBUG_RETURN(result);
+}
```
Comment (midenok) [inline sql/ha_partition.cc]: `loop_extra()` is empty stub. Use `loop_partitions()` everywhere.

**Review by midenok — COMMENTED** [review]
Message typo:

> * update keyread state of underlining handlers recursively

_underlying_

File: sql/ha_partition.cc [hunk 3]
```diff
@@ -9075,23 +9104,23 @@ int ha_partition::loop_extra_alter(enum ha_extra_function operation)
...
+    @param f                     callback
+    @param param                 a void*-parameter passed to callback
+
+    @return Operation status
+      @retval >0                    Error code
+      @retval 0                     Success
 */
 
-int ha_partition::loop_extra(enum ha_extra_function operation)
+int ha_partition::loop_partitions(int f(handler *, void *), void *param)
```
Comment (midenok) [inline sql/ha_partition.cc]: I would name it **callback**. Easier to apprehend.

File: sql/ha_partition.cc (hunk 2 above)
Comment (FooBarrior) [reply sql/ha_partition.cc]: `loop_extra = loop_partitions(extra_cb, .)`
Using `loop_partitions` directly would increase patch significantly. Let's assume it as a shortcut.
Maybe I'd better just remove debug frame and make it inline?

Comment (midenok) [reply sql/ha_partition.cc]: I looked at how much there `loop_extra()` references before writing the recommendation. It's not much.

**Pull request merged**